[WIP] Added method cuda to PackedSequence.

jusjusjus opened the pull request.

jusjusjus committed:
Imputed `torch.tensor` methods to `PackedSequence`

The tensor transformations `cuda`, `cpu`, `half`, and `double` are
imputed into class `PackedSequence`, as convenience functions to
transform its `data` attribute.

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -26,7 +26,8 @@ class PackedSequence(PackedSequence_):
     """
-    pass
+    def cuda(self):
+        return type(self)(self.data.cuda(), self.batch_sizes)
```
Can you add `cpu`, `float`, `double` and `half` too? Might be a good idea to implement them like this:

```python
def _make_cast(name):
  def fn(self, *args, **kwargs):
    return type(self)(getattr(self.data, name)(*args, **kwargs), self.batch_sizes)
  return fn

cuda = _make_cast('cuda')
cpu = _make_cast('cpu')
```

jusjusjus replied:
Good one, I didn't think of the cuda args.  I really like the syntax, however:
```
import time
num = 10**7

class foo:
    bar = 5

f = foo()

t0 = time.time()
[f.bar  for _ in range(num)]
print('f.bar :', (time.time()-t0)/num)

t0 = time.time()
[getattr(f, 'bar') for _ in range(num)]
print('getattr(f,"bar") :', (time.time()-t0)/num)

> f.bar : 5.444774627685547e-08
> getattr(f,"bar") : 1.235971450805664e-07
```
Questions:
(1) Do you think someone will want to call this more than once per batch?
(2) Also what do you think of a generalization `_gen_cast_of(attr, name)`?

apaszke replied:
I don't think the difference matters this much in here, but I'm fine with repeating all the cases manually too. Can you explain what do you mean in (2)?

jusjusjus commented:
Ok, I made the changes, squashed and rebased to master.  Maybe `impute` is not the right word here.

I took care that the call preserves class inheritance by `type(self)(...)`.

jusjusjus commented:
Ok, I made the changes, squashed and rebased to master. Maybe impute is not the right word here.

I took care that the call preserves class inheritance by type(self)(...).

apaszke commented:
Oh actually can you please add some tests for it?

jusjusjus committed:
Added Prop `is_cuda` and fixed bug.

* `tensor.is_cuda` property is imputed to `PackedSequence`.

* Bug in argument `padding_value` of `pad_packed_sequence` is fixed.
For packed tensors of non-float types, `pad_packed_sequence` failed
because `padding_value` has to be `0` (not `0.0`).

jusjusjus commented:
There was another bug that I addressed in `pad_packed_sequence` with the padding value being of type float.  I'll try to add tests next, but maybe I could use some guidance here :)

apaszke commented:
You need to add a method somewhere in `test/test_nn.py` that checks if the casts are successful.

jusjusjus committed:
Added test for `PackedSequence` masks.

jusjusjus commented:
I created a `PackedSequenceTest` class to test utils/rnn.py stuff.  If that's good, I'd merge the other `rnn.py` stuff into that class as well.  Should we run through different types for these tests, as this led to errors in `pad_packed_sequence` before?

Also, somethings up with `HalfTensor`.   It doesn't have a method `fill_`.

I designed the test without reference to `packed.data` because I think an attribute `data` could be left flexible in the long run, and its presence/disappearance shouldn't break test.

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -82,7 +104,19 @@ def pack_padded_sequence(input, lengths, batch_first=False):
+    'torch.ShortTensor': 0,
+    'torch.CharTensor': 0,
+    'torch.ByteTensor': 0,
+}
```
Is that really necessary? It seemed to be working fine before. A better way would be to use `torch._integer_tensor_classes` (the map you wrote here doesn't handle e.g. CUDA and sparse types).

jusjusjus replied:
Yeah, except for int types:
```import torch
import numpy as np
from torch.autograd import Variable
batch_size = 12
max_length = 4
sequences = [
    Variable(torch.rand(np.random.randint(1, max_length+1)))
    for _ in range(batch_size)
]
ordered = sorted(sequences, key=len, reverse=True)
lengths = [len(x) for x in ordered]
padded = torch.nn.utils.rnn.pad_sequence(ordered)
padded = padded.int()
packed = torch.nn.utils.rnn.pack_padded_sequence(padded, lengths)
torch.nn.utils.rnn.pad_packed_sequence(packed)
```

yields `TypeError: fill_ received an invalid combination of arguments - got (float), but expected (int value)` in origin/master and python 3.5.

jusjusjus replied:
check

apaszke commented on test/test_nn.py:
```diff
@@ -36,6 +36,65 @@
+        self.batch_size = 5
+        self.max_length = 6
+
+    def _ordered_sequence(self, type):
```
Please don't shadow the `type` builtin. It took me a moment to figure out why this code doesn't crash

jusjusjus replied:
check

apaszke commented on test/test_nn.py:
```diff
@@ -36,6 +36,65 @@
+        'torch.FloatTensor': (torch.FloatTensor, 'float'),
+        # TODO 'torch.HalfTensor': (torch.HalfTensor, 'half'),
+        # * First need to fix:
+        # AttributeError: 'torch.HalfTensor' object has no attribute 'fill_'
```
Known thing, won't fix.

apaszke commented on test/test_nn.py:
```diff
@@ -36,6 +36,65 @@
+        padded_tensor = rnn_utils.pad_sequence(ordered)
+        return padded_tensor, lengths
+
+    def test_type_masks(self):
```
can you change `mask` to `cast`? We usually call these operations "type casts". Masking is when you use a boolean tensor to select/change/copy only a subset of values.

jusjusjus replied:
check

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -26,7 +26,29 @@ class PackedSequence(PackedSequence_):
+        setattr(cls, name, fn)
+
+# Impute following `torch.Tensor` methods and properties into `PackedSequence`
+_methods_to_impute = 'cuda cpu double float long int short char byte'.split()
```
It would be simpler and more explicit to just write them out as a list

jusjusjus replied:
Sure, we could still impute all the different cos/acos/sin/sin_ etc. methods, or should we write a wrapper for each one?

jusjusjus committed:
Addressed all amendments of apaszke.

apaszke commented on torch/nn/utils/rnn.py:
```diff
@@ -82,7 +131,7 @@ def pack_padded_sequence(input, lengths, batch_first=False):
 
 
-def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
+def pad_packed_sequence(sequence, batch_first=False, padding_value=None):
```
Actually wouldn't changing the default to `0` (an int) suffice?

jusjusjus committed:
Fix bug by defaulting to .

jusjusjus commented:
You're right.  I just tested it under py3.5.

2018-01-09 20:54 GMT+01:00 Adam Paszke <notifications@github.com>:

> *@apaszke* commented on this pull request.
> ------------------------------
>
> In torch/nn/utils/rnn.py
> <https://github.com/pytorch/pytorch/pull/4430#discussion_r160508764>:
>
> > @@ -82,7 +131,7 @@ def pack_padded_sequence(input, lengths, batch_first=False):
>      return PackedSequence(torch.cat(steps), batch_sizes)
>
>
> -def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0):
> +def pad_packed_sequence(sequence, batch_first=False, padding_value=None):
>
> Actually wouldn't changing the default to 0 (an int) suffice?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/4430#pullrequestreview-87638875>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGAdFDYoZnLk-aqWPfV9yAzzd5m2Isagks5tI8PogaJpZM4RQIXA>
> .
>



-- 
Justus Schwabedal

Handy (D): +49 177 939 5281
Handy (US): +1 857 919 1430
email: jschwabedal@googlemail.com
skype: justus1802

144 Moreland Ave NE, Apt 524
Atlanta, GA, 30307
USA

Steinkreuzstr. 21
53757 Sankt Augustin, NRW
Germany

jusjusjus committed:
Reverted doc-change in pad_packed_sequence

apaszke commented:
@pytorchbot test this please

apaszke commented:
Thank you!

jusjusjus commented:
Thanks for your guidance, @apaszke.
